include/cantera/numerics/FuncEval.h

```diff
@@ -137,6 +137,19 @@ class FuncEval
      */
     int preconditioner_solve_nothrow(double* rhs, double* output);
 
+    //! Number of event/root functions exposed to the integrator (0 disables root finding)
+    virtual size_t nRootFunctions() const {
+        return 0;
+    }
+
+    /** Evaluate the event/root functions currently in play.
+     * Integrators invoke this whenever root finding is enabled; implementations
+     * should fill `gout` with the function values and return 0 on success.
+     */
+    virtual int evalRootFunctions(double t, const double* y, double* gout) {
+        return 0;
+    }
```

Comment:
Since this function is called by CVODES, I think it also needs to have the protection against throwing exceptions that we implement for the other callback methods, for example the virtual `eval` method which is called by the non-virtual `evalNoThrow` method, where the latter is the one that is set as the CVODES callback.

Also, as a minor nitpick, I would prefer formatting the docstring as:

```suggestion
    /**
     * Evaluate the event/root functions currently in play.
     * Integrators invoke this whenever root finding is enabled; implementations
     * should fill `gout` with the function values and return 0 on success.
     */
    virtual int evalRootFunctions(double t, const double* y, double* gout) {
        return 0;
    }
```
i.e., avoid starting the comment on the `/**` line. If you want a more compact docstring, I'd suggest the `//!` syntax.